io/proc_reader.cc

```diff
@@ -87,6 +87,166 @@ size_t find_nth(string_view str, char c, uint32_t index) {
+bool ParseSocketLine(string_view line, ino_t target_inode, bool is_ipv6, TcpInfo* info) {
+  line = absl::StripLeadingAsciiWhitespace(line);
+
+  std::vector<string_view> parts;
+  size_t pos = 0;
+
+  while (pos < line.size()) {
+    while (pos < line.size() && absl::ascii_isspace(line[pos]))
+      pos++;
+    if (pos >= line.size())
+      break;
+
+    size_t token_start = pos;
+    while (pos < line.size() && !absl::ascii_isspace(line[pos]))
+      pos++;
+
+    parts.push_back(line.substr(token_start, pos - token_start));
+  }
+
+  if (parts.size() < 10)
+    return false;
+
+  unsigned int sl;
+  string_view sl_part = parts[0];
+  if (absl::EndsWith(sl_part, ":")) {
+    sl_part.remove_suffix(1);
+  }
+  if (!absl::SimpleAtoi(sl_part, &sl))
+    return false;
+
+  string_view local_addr_port = parts[1];
+  size_t colon_pos = local_addr_port.find(':');
+  if (colon_pos == string_view::npos)
+    return false;
+
+  string_view local_addr_hex = local_addr_port.substr(0, colon_pos);
+  string_view local_port_hex = local_addr_port.substr(colon_pos + 1);
+
+  string_view remote_addr_port = parts[2];
+  colon_pos = remote_addr_port.find(':');
+  if (colon_pos == string_view::npos)
+    return false;
+
+  string_view remote_addr_hex = remote_addr_port.substr(0, colon_pos);
+  string_view remote_port_hex = remote_addr_port.substr(colon_pos + 1);
+
+  unsigned int state;
+  if (!absl::SimpleHexAtoi(parts[3], &state))
+    return false;
+
+  unsigned int inode = 0;
+  if (!absl::SimpleAtoi(parts[9], &inode))
+    return false;
+
+  if (inode != target_inode)
+    return false;
+
+  info->is_ipv6 = is_ipv6;
+  info->state = state;
+  info->inode = inode;
+
+  unsigned int port;
+  if (absl::SimpleHexAtoi(local_port_hex, &port)) {
+    info->local_port = port;
+  }
+  if (absl::SimpleHexAtoi(remote_port_hex, &port)) {
+    info->remote_port = port;
+  }
+
+  if (is_ipv6) {
+    HexToIPv6(local_addr_hex, info->local_addr6);
+    HexToIPv6(remote_addr_hex, info->remote_addr6);
+  } else {
+    unsigned int addr;
+    if (absl::SimpleHexAtoi(local_addr_hex, &addr)) {
+      uint8_t* bytes = reinterpret_cast<uint8_t*>(&addr);
+      info->local_addr = (bytes[0] << 24) | (bytes[1] << 16) | (bytes[2] << 8) | bytes[3];
+    }
+    if (absl::SimpleHexAtoi(remote_addr_hex, &addr)) {
+      uint8_t* bytes = reinterpret_cast<uint8_t*>(&addr);
+      info->remote_addr = (bytes[0] << 24) | (bytes[1] << 16) | (bytes[2] << 8) | bytes[3];
+    }
+  }
+
+  return true;
+}
+
+// Reads TCP info from a specified proc file
+Result<TcpInfo> ReadTcpInfoFromFile(const char* proc_path, ino_t sock_inode, bool is_ipv6) {
+  TcpInfo info;
+  int fd = open(proc_path, O_RDONLY | O_CLOEXEC);
+  if (fd < 0) {
+    return make_unexpected(error_code{errno, system_category()});
+  }
+
+  base::IoBuf buf{4096};
+  bool found = false;
+  bool header_skipped = false;
+
+  while (true) {
+    auto dest = buf.AppendBuffer();
+
+    int res = read(fd, dest.data(), dest.size());
```

Comment:
Helio has utilities to read and write to a file -- you can use these and work well with the async interface of iouring. As is, we have a `blocking` call, meaning that it blocks the thread until the operation is done. Furthermore, the open without the O_DIRECT goes through the page cache. What we get here is buffered IO.

Not sure if we are fine with these blocking calls. Maybe @romange can add some comment here

Reply:
plz see `io/file.h` for how we do it inhelio

Reply:
it's not needed here, as `/proc/` filesystem is special.